devicedefender/source/DeviceDefender.cpp

```diff
@@ -114,6 +190,189 @@ namespace Aws
+            aws_iotdevice_defender_config_register_ip_list_metric(
+                m_taskConfig, &cursor, CustomMetricIpList::GetMetricFunction, data.get());
+        }
+
+        void ReportTask::RegisterCustomMetricCpuUsage() noexcept
+        {
+            CustomMetricNumberFunction func =
+                std::bind(&ReportTask::getCustomMetricCpuUsage, this, std::placeholders::_1);
+            RegisterCustomMetricNumber("cpu_usage", std::move(func));
+        }
+
+        int ReportTask::getCustomMetricCpuUsage(double *output)
+        {
+// Get the CPU usage from Linux
+#if defined(__linux__) || defined(__unix__)
+            int return_result = AWS_OP_ERR;
+            uint64_t totalUser, totalUserLow, totalSystem, totalIdle, total;
+            getCurrentCpuUsage(&totalUser, &totalUserLow, &totalSystem, &totalIdle);
+            double percent;
+
+            // Overflow detection
+            if (totalUser < m_cpuLastTotalUser || totalUserLow < m_cpuLastTotalUserLow ||
+                totalSystem < m_cpuLastTotalSystem || totalIdle < m_cpuLastTotalIdle)
+            {
+                *output = 0;
+            }
+            else
+            {
+                total = (totalUser - m_cpuLastTotalUser) + (totalUserLow - m_cpuLastTotalUserLow) +
+                        (totalSystem - m_cpuLastTotalSystem);
+                percent = total;
+                total += totalIdle - m_cpuLastTotalIdle;
+                percent = (percent / total) * 100;
+
+                // If percent is negative, then there was an error (overflow?)
+                if (percent < 0)
+                {
+                    *output = 0;
+                    return_result = AWS_OP_ERR;
+                }
+                else
+                {
+                    *output = percent;
+                    return_result = AWS_OP_SUCCESS;
+                }
+            }
+
+            m_cpuLastTotalUser = totalUser;
+            m_cpuLastTotalUserLow = totalUserLow;
+            m_cpuLastTotalSystem = totalSystem;
+            m_cpuLastTotalIdle = totalIdle;
+
+            return return_result;
+#endif
+
+            // OS not supported? Just return an error and set the output to 0
+            *output = 0;
+            return AWS_OP_ERR;
+        }
+
+        void ReportTask::getCurrentCpuUsage(
+            uint64_t *totalUser,
+            uint64_t *totalUserLow,
+            uint64_t *totalSystem,
+            uint64_t *totalIdle)
+        {
+            *totalUser = 0;    // prevent warnings over unused parameter on Windows and Mac
+            *totalUserLow = 0; // prevent warnings over unused parameter on Windows and Mac
+            *totalSystem = 0;  // prevent warnings over unused parameter on Windows and Mac
+            *totalIdle = 0;    // prevent warnings over unused parameter on Windows and Mac
+
+// Get the CPU usage from Linux
+#if defined(__linux__) || defined(__unix__)
```

Comment:
Ideally, this should be moved into an aws-c-common or aws-c-io API with per-platform implementations